rfcs/20221006-frontend_event_tracking.md

```diff
@@ -0,0 +1,296 @@
+    provider: string;
+    /** The search term **/
+    query: string;
+  };
+  /** The user clicks the CTA button to the external source to use the image */
+  GET_MEDIA: {
+    /** The unique ID of the media */
+    id: string;
+    /** The slug (not the prettified name) of the provider */
+    provider: string;
+    /** The media type being searched */
+    mediaType: MediaType;
+  };
+  /** The user clicks one of the buttons to copy the media attribution */
+  COPY_ATTRIBUTION: {
+    /** The unique ID of the media */
+    id: string;
+    /** The format of the copied attribution */
+    format: "plain" | "rich" | "html";
+    /** The media type being searched */
+    mediaType: MediaType;
+  };
+  /** The user reports a piece of media through our form */
+  REPORT_MEDIA: {
+    /** The unique ID of the media */
+    id: string;
+    /** The slug (not the prettified name) of the provider */
+    provider: string;
+    /** The media type being searched */
+    mediaType: MediaType;
+  };
+  /** The user plays, pauses, or seeks an audio track.
+   * @todo: This potentially requires throttling.
+   **/
+  AUDIO_INTERACTION: {
+    /** The unique ID of the media */
+    id: string;
+    event: "seek" | "play" | "pause";
+    /** The slug (not the prettified name) of the provider */
+    provider: string;
+  };
+  /** The user visits a CC license description page on CC.org  **/
+  VISIT_LICENSE_PAGE: {
+    /** The slug of the license the user clicked on */
+    license: string;
+  };
+  /** The user visits a creator's link in the single result UI  **/
+  VISIT_CREATOR_LINK: {
+    /** The unique ID of the media */
+    id: string;
+    /** The permalink to the creator's profile */
+    url: string;
+  };
+  /** The user right clicks a single image result, most likely to download it. **/
+  RIGHT_CLICK_IMAGE: {
+    /** The unique ID of the media */
+    id: string;
+    /** The media type being searched */
+    mediaType: MediaType;
+  };
+  /** The user uses the 'back to search' link on a single result **/
+  BACK_TO_SEARCH: {
+    /** The unique ID of the media */
+    id: string;
+    /** The media type being searched */
+    mediaType: MediaType;
+  };
+  /** The visibility of the filter sidebar on desktop is toggled **/
+  TOGGLE_FILTERS: {
+    /** The media type being searched */
+    mediaType: MediaType;
+  };
+};
+type AnalyticsEventTypes = keyof AnalyticsEvents;
+
+/**
+ * Send an event to our analytics endpoint. Should be a no-op if the
+ * user has opted-out of analytics.
+ */
+async function sendEvent<T extends AnalyticsEventTypes>(
+  key: T,
+  payload?: AnalyticsEvents[T]
+): Promise<any> {
+  /**
+   * The default values to send with every analytics event.
+   */
+  const defaults = {
+    timestamp: Date.now(),
+    path: window.location.pathname,
+    // The dimensions of the user's device
+    width: window.innerWidth,
+    height: window.innerHeight,
+    referrer: document.referrer || null,
```

Comment:
I am not sure if collecting the following information is invasive but it would be beneficial to have information about the browser (name and version), the operating system (name and version) and the device. We typically ask for this information in our bug reports.

Reply:
I'm not even sure that specific width and height are necessary for the questions we currently have and would like us to consider collecting _as little_ information as possible. The information that might be useful for us is the same kind of stuff in the UI state cookie, general breakpoint (maybe?, I'm sceptical...) and whether it's a mobile UA (definitely useful considering the significant design differences). For bugs, we'd ideally want to rely on Sentry for that I would think?

Reply:
I didn't mean to suggest using analytics as bug reports, the reference to the issue template was just to say that there is precedent for asking for this info. While on one hand, I can get behind collecting as coarse information as possible, but on the other, it might help to how certain interactions are affected by the devices in use.

Reply:
That's true, device capability (low-spec devices for example) _could_ be useful, I didn't consider that. Some of this information will not be easily derived from flattened UA string we have now though and some (like RAM and CPU capacity) will require permission prompts when UA Client Hints are rolled out: https://developer.mozilla.org/en-US/docs/Web/API/User-Agent_Client_Hints_API

We can reliably get browser and device type (I think) without permission prompts.

Reply:
That sounds good. The derived data will probably be a _lot_ easier (and faster) to query as well.

Reply:
Fields and documentation added in 34fea2a57093c38c25ae218a8fc4ac197d3bea6d.